File: FlagMiner/ImportForm.vb
Imports System.Xml

Public Class ImportForm

    Public links As List(Of Tuple(Of String, String, String))

    Private Sub Button1_Click(sender As Object, e As EventArgs) Handles Button1.Click
        Dim errorCode As Integer
        Dim rawResponse As String
        'Dim board, thread As String

        Dim uri As New Uri(TextBox1.Text)
        Dim parsedUrl As String = uri.GetLeftPart(UriPartial.Path)

        errorCode = Form1.loadThread(Nothing, Nothing, rawResponse, parsedUrl)

        Dim posts As List(Of Post)
        errorCode = Form1.parseThread(rawResponse, posts)

        Dim sourcePosts As New List(Of ULong)
        errorCode = parseStrings(posts, sourcePosts)

        Dim trimmedPosts As New List(Of Post)
        errorCode = filterPosts(posts, sourcePosts, trimmedPosts)

        links = New List(Of Tuple(Of String, String, String))
        errorCode = gatherLinks(trimmedPosts, links)

        Me.DialogResult = Windows.Forms.DialogResult.OK
    End Sub

    Private Sub Button2_Click(sender As Object, e As EventArgs) Handles Button2.Click
        Me.DialogResult = Windows.Forms.DialogResult.Cancel
    End Sub

    Public Function parseStrings(ByRef posts As List(Of Post), ByRef sourcePosts As List(Of ULong)) As Integer
        For Each lin As String In TextBox2.Lines
            Dim temp As String = lin.Trim({">"c, " "c})
            Dim res As ULong
            If String.IsNullOrEmpty(temp) Then
                Continue For
            End If
            If ULong.TryParse(temp, res) Then
                sourcePosts.Add(res)
            Else
                MsgBox("Wrong format for " & lin, , "Flag Miner")
                Return 1
            End If
        Next
        sourcePosts.Distinct()
        sourcePosts.Sort()

        Return 0
    End Function

    Public Function filterPosts(ByRef posts As List(Of Post), ByRef sourcePosts As List(Of ULong), filtered As List(Of Post)) As Integer
        For Each Post As Post In posts
            If sourcePosts.Contains(Post.no) Then
                filtered.Add(Post)
            End If
        Next
        Return 0
    End Function


    Public Function gatherLinks(ByRef filtered As List(Of Post), ByRef links As List(Of Tuple(Of String, String, String))) As Integer
        Dim temp As New List(Of String)

        WebBrowser1.Navigate(String.Empty)
        Dim fakeDoc As HtmlDocument = WebBrowser1.Document
        For Each Post As Post In filtered
            Dim text As String = Post.com
            text = "<HTML><body>" & text & "</body></HTML>"

            fakeDoc.Write(text)
        Next

        Dim HtmlElems As HtmlElementCollection = fakeDoc.Links
        For Each HtmlElem As HtmlElement In HtmlElems
            Dim str As String = HtmlElem.GetAttribute("href")
            Dim uri As New Uri(str)
            Dim str2 As String = uri.Fragment
            Dim parts As String() = uri.AbsolutePath.Split({"/"c}, StringSplitOptions.RemoveEmptyEntries)
            If String.IsNullOrEmpty(str2) Then
                str2 = parts(2)
            Else
                str2 = str2.Trim({"#"c, "p"c})
            End If
            If parts.Length = 3 Then links.Add(New Tuple(Of String, String, String)(parts(0), parts(2), str2))
        Next
        Return 0

    End Function


    Private Sub ImportForm_Load(sender As Object, e As EventArgs) Handles MyBase.HelpRequested
        WatDo.ShowDialog()
    End Sub
End Class

File: FlagMiner/ImportForm.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;
using System.Linq;
using System.Xml.Linq;
using System.Threading.Tasks;
using System.Xml;
using System.Net;

namespace FlagMiner
{

	partial class ImportForm :Form
	{
        public Form1 myForm1;
        public ImportForm(Form1 frm1)
        {
            myForm1 = frm1;
            InitializeComponent();
        }

		public List<Tuple<string, string, string>> links;
		private void Button1_Click(object sender, EventArgs e)
		{
            try
            {
                int statusFlag = 0;
                int errorCode = 0;
                string rawResponse = null;
                //Dim board, thread As String

                Uri uri = new Uri(TextBox1.Text);
                string parsedUrl = uri.GetLeftPart(UriPartial.Path);

                errorCode = myForm1.LoadThread(null, 0, out rawResponse, parsedUrl);
                Form1.RaiseError(errorCode, ref statusFlag);

                List<Post> posts = null;
                myForm1.ParseThread(rawResponse, ref posts);

                List<ulong> sourcePosts = new List<ulong>();
                errorCode = parseStrings(ref posts, ref sourcePosts);
                Form1.RaiseError(errorCode, ref statusFlag);

                List<Post> trimmedPosts = new List<Post>();
                filterPosts(ref posts, ref sourcePosts, trimmedPosts);

                links = new List<Tuple<string, string, string>>();
                gatherLinks(ref trimmedPosts, ref links);

                this.DialogResult = System.Windows.Forms.DialogResult.OK;
            }
            catch (WebException ex)
            {
                myForm1.AppendText(DateTime.Now + " : " + ex.ToString() + System.Environment.NewLine);
            }
            catch (Exception ex)
            {
                myForm1.AppendText(DateTime.Now + " : " + ex.ToString() + System.Environment.NewLine);
            }
		}

        private void Button2_Click(object sender, EventArgs e) => this.DialogResult = System.Windows.Forms.DialogResult.Cancel;

        public int parseStrings(ref List<Post> posts, ref List<ulong> sourcePosts)
		{
			foreach (string lin in TextBox2.Lines) {
				string temp = lin.Trim( new char[] {
					'>',
					' '
				});
				ulong res = 0;
				if (string.IsNullOrEmpty(temp)) {
					continue;
				}
				if (ulong.TryParse(temp, out res)) {
					sourcePosts.Add(res);
				} else {
                    MessageBox.Show("Wrong format for " + lin, "Flag Miner", MessageBoxButtons.OK, MessageBoxIcon.Error);
					return 1;
				}
			}
			sourcePosts = (List<ulong>)sourcePosts.Distinct();
			sourcePosts.Sort();

			return 0;
		}

		public void filterPosts(ref List<Post> posts, ref List<ulong> sourcePosts, List<Post> filtered)
		{
			foreach (Post Post in posts) {
				if (sourcePosts.Contains((ulong)Post.no)) {
					filtered.Add(Post);
				}
			}
		}


		public void gatherLinks(ref List<Post> filtered, ref List<Tuple<string, string, string>> links)
		{

			WebBrowser1.Navigate(string.Empty);
			HtmlDocument fakeDoc = WebBrowser1.Document;
			foreach (Post Post in filtered) {
				string text = Post.com;
				text = "<HTML><body>" + text + "</body></HTML>";

				fakeDoc.Write(text);
			}

			HtmlElementCollection HtmlElems = fakeDoc.Links;
			foreach (HtmlElement HtmlElem in HtmlElems) {
				string str = HtmlElem.GetAttribute("href");
				Uri uri = new Uri(str);
				string str2 = uri.Fragment;
				string[] parts = uri.AbsolutePath.Split(new char[] { '/' }, StringSplitOptions.RemoveEmptyEntries);
				if (string.IsNullOrEmpty(str2)) {
					str2 = parts[2];
				} else {
					str2 = str2.Trim(new char[] {
						'#',
						'p'
					});
				}
				if (parts.Length == 3)
					links.Add(new Tuple<string, string, string>(parts[0], parts[2], str2));
			}
		}

        WatDo watDo=null;
		private void ImportForm_HelpRequested(object sender, EventArgs e)
		{
            if (watDo == null) watDo = new WatDo();
			watDo.ShowDialog();
		}
	}
}
